### PR TITLE
Check DB version at startup

### LINKED ISSUE #1: Check DB version at startup
Forgot to add both the store and the check at startup

### PR DESCRIPTION
**Pull request by danielabrozzoni**
Closes #1

### CODE REVIEW

File: src/db/mod.rs [hunk 1]
```diff
@@ -52,10 +60,23 @@ async fn establish_connection(
...
+        let statement = client
+            .prepare_typed("INSERT INTO version(version) VALUES ($1)", &[Type::INT4])
+            .await?;
+        client.execute(&statement, &[&(DB_VERSION as i32)]).await?;
+    } else {
+        let db_version = db_version[0].get::<_, i32>("version") as u32;
+        if db_version != DB_VERSION {
+            return Err(DbError::WrongVersion(db_version));
+        }
+    }
```
Comment (darosior) [inline src/db/mod.rs:79]: How about using `query_opt`? Seems more adapted to me:
```diff
diff --git a/src/db/mod.rs b/src/db/mod.rs
index 2d48c37..b141260 100644
--- a/src/db/mod.rs
+++ b/src/db/mod.rs
@@ -13,7 +13,7 @@ use std::{collections::BTreeMap, fmt};
 
 use tokio_postgres::{types::Type, Client, NoTls};
 
-pub const DB_VERSION: u32 = 0;
+pub const DB_VERSION: i32 = 0;
 
 #[derive(Debug)]
 pub enum DbError {
@@ -21,7 +21,7 @@ pub enum DbError {
     Postgres(tokio_postgres::Error),
     /// Trying to insert the same data twice
     Duplicate,
-    WrongVersion(u32),
+    WrongVersion(i32),
 }
 
 impl fmt::Display for DbError {
@@ -64,18 +64,18 @@ pub async fn maybe_create_db(config: &tokio_postgres::Config) -> Result<(), DbEr
     let client = establish_connection(config).await?;
 
     client.batch_execute(SCHEMA).await?;
-    let db_version = client.query("SELECT version FROM VERSION", &[]).await?;
-    if db_version.is_empty() {
+    let db_version = client.query_opt("SELECT version FROM VERSION", &[]).await?;
+    if let Some(db_ver) = db_version {
+        let db_ver = db_ver.get(0);
+        if db_ver != DB_VERSION {
+            return Err(DbError::WrongVersion(db_ver));
+        }
+    } else {
         // Alright, we just created the db. Let's insert the version...
         let statement = client
             .prepare_typed("INSERT INTO version(version) VALUES ($1)", &[Type::INT4])
             .await?;
         client.execute(&statement, &[&(DB_VERSION as i32)]).await?;
-    } else {
-        let db_version = db_version[0].get::<_, i32>("version") as u32;
-        if db_version != DB_VERSION {
-            return Err(DbError::WrongVersion(db_version));
-        }
     }
 
     Ok(())
```

**Review by darosior — COMMENTED** [review]
Cool. See how the codebase is messy now? :/

Here is a test for the DB version check:
```diff
@@ -218,3 +218,131 @@ pub async fn fetch_spend_tx(
 
     Ok(spend_tx.map(|tx| encode::deserialize(&tx).expect("Added to DB with serialize()")))
 }
+
+#[cfg(test)]
+mod tests {
+    use crate::db::*;
+
+    use revault_net::{
+        bitcoin::{
+            hashes::hex::FromHex,
+            secp256k1::{PublicKey, Signature},
+            OutPoint, Txid,
+        },
+        message::{coordinator::*, ResponseResult},
+    };
+    use revault_tx::transactions::{RevaultTransaction, SpendTransaction};
+
+    use std::{collections::BTreeMap, str::FromStr};
+
+    use tokio::runtime::Builder as RuntimeBuilder;
+    use tokio_postgres::tls::NoTls;
+
+    async fn create_test_db() {
+        let conf =
+            tokio_postgres::Config::from_str("postgresql://revault:revault@localhost/postgres")
+                .unwrap();
+
+        let (client, connection) = conf.connect(NoTls).await.unwrap();
+        tokio::spawn(async move {
+            if let Err(e) = connection.await {
+                log::error!("Database connection error: {}", e);
+            }
+        });
+
+        client
+            .batch_execute("DROP DATABASE IF EXISTS coordinatord_test_db;")
+            .await
+            .expect("dropping tables");
+        client
+            .batch_execute("CREATE DATABASE coordinatord_test_db;")
+            .await
+            .expect("creating tables");
+    }
+
+    async fn postgre_setup() -> tokio_postgres::Config {
+        create_test_db().await;
+        let conf = tokio_postgres::Config::from_str(
+            "postgresql://revault:revault@localhost/coordinatord_test_db",
+        )
+        .unwrap();
+
+        // Cleanup any leftover
+        let (client, connection) = conf.connect(NoTls).await.unwrap();
+        tokio::spawn(async move {
+            if let Err(e) = connection.await {
+                log::error!("Database connection error: {}", e);
+            }
+        });
+        client
+            .batch_execute("DROP TABLE IF EXISTS signatures; DROP TABLE IF EXISTS spend_outpoints; DROP TABLE IF EXISTS spend_txs; DROP TABLE IF EXISTS version;")
+            .await
+            .expect("dropping tables");
+
+        conf
+    }
+
+    async fn postgre_teardown(conf: &tokio_postgres::Config) {
+        let (client, connection) = conf.connect(NoTls).await.unwrap();
+        tokio::spawn(async move {
+            if let Err(e) = connection.await {
+                log::error!("Database connection error: {}", e);
+            }
+        });
+        client
+            .batch_execute("DROP TABLE signatures; DROP TABLE spend_outpoints; DROP TABLE spend_txs; DROP TABLE version;")
+            .await
+            .expect("dropping tables");
+    }
+
+    #[test]
+    fn db_version_check() {
+        let rt = RuntimeBuilder::new_multi_thread()
+            .enable_all()
+            .thread_name("coordinatord_test_db")
+            .build()
+            .expect("Creating tokio runtime");
+
+        async fn run_test() {
+            let pg_config = postgre_setup().await;
+            let client = establish_connection(&pg_config).await.unwrap();
+
+            // The first call will create the DB
+            assert!(
+                client
+                    .query_opt("SELECT version FROM VERSION", &[])
+                    .await
+                    .is_err(),
+                "This relation shouldn't exist yet"
+            );
+            maybe_create_db(&pg_config).await.unwrap();
+            assert!(
+                client
+                    .query_opt("SELECT version FROM VERSION", &[])
+                    .await
+                    .unwrap()
+                    .is_some(),
+                "It should have been populated already!"
+            );
+
+            // The second one will sanity check it's right
+            maybe_create_db(&pg_config).await.unwrap();
+
+            // We'll refuse to start if the version is from the future
+            let statement = client
+                .prepare_typed("UPDATE version SET version = 1", &[])
+                .await
+                .unwrap();
+            client.execute(&statement, &[]).await.unwrap();
+            assert!(maybe_create_db(&pg_config)
+                .await
+                .unwrap_err()
+                .to_string()
+                .contains("Unexpected database version"));
+
+            postgre_teardown(&pg_config).await;
+        }
+
+        rt.block_on(run_test());
+    }
+}
```

File: src/db/mod.rs (hunk 1 above)
Comment (danielabrozzoni) [reply src/db/mod.rs:79]: It's way cleaner, you're right :)

File: src/db/mod.rs [hunk 2]
```diff
@@ -52,10 +60,23 @@ async fn establish_connection(
...
+        let db_ver = db_ver.get(0);
+        if db_ver != DB_VERSION {
+            return Err(DbError::WrongVersion(db_ver));
+        }
+    } else {
+        // Alright, we just created the db. Let's insert the version...
+        let statement = client
+            .prepare_typed("INSERT INTO version(version) VALUES ($1)", &[Type::INT4])
+            .await?;
+        client.execute(&statement, &[&(DB_VERSION as i32)]).await?;
```
Comment (darosior) [inline src/db/mod.rs:78]: It's already an `i32`?

**Review by darosior — COMMENTED** [review]
Just a nit

**Comment by darosior** [comment]
ACK 574070acbd5c9cbfa37fba0a2457bce1a8bc43eb

Unfortunately i didn't merge this before the testing so it'll go post release.

**Pull request merged**